README.md
# Target Redshift

[![CircleCI](https://circleci.com/gh/datamill-co/target-redshift.svg?style=svg)](https://circleci.com/gh/datamill-co/target-redshift)

[![PyPI version](https://badge.fury.io/py/target-redshift.svg)](https://pypi.org/project/target-redshift/)

[![](https://img.shields.io/librariesio/github/datamill-co/target-redshift.svg)](https://libraries.io/github/datamill-co/target-redshift)

A [Singer](https://singer.io/) redshift target, for use with Singer streams generated by Singer taps.

## Features

- Creates SQL tables for [Singer](https://singer.io) streams
- Denests objects flattening them into the parent object's table
- Denests rows into separate tables
- Adds columns and sub-tables as new fields are added to the stream [JSON Schema](https://json-schema.org/)
- Full stream replication via record `version` and `ACTIVATE_VERSION` messages.

## Install

```sh
pip install target-redshift
```

## Usage

1. Follow the
   [Singer.io Best Practices](https://github.com/singer-io/getting-started/blob/master/docs/RUNNING_AND_DEVELOPING.md#running-a-singer-tap-with-a-singer-target)
   for setting up separate `tap` and `target` virtualenvs to avoid version
   conflicts.

1. Create a [config file](#configjson) at
   `~/singer.io/target_redshift_config.json` with redshift connection
   information and target redshift schema.

   ```json
   {
     "redshift_host": "aws.something.or.other",
     "redshift_port": 5439,
     "redshift_database": "my_analytics",
     "redshift_username": "myuser",
     "redshift_password": "1234",
     "redshift_schema": "mytapname",
     "default_column_length": 1000,
     "default_max_bytes_per_character": 2,
     "target_s3": {
       "aws_access_key_id": "AKIA...",
       "aws_secret_access_key": "supersecret",
       "bucket": "target_redshift_staging",
       "key_prefix": "__tmp"
     }
   }
   ```

1. Run `target-redshift` against a [Singer](https://singer.io) tap.

   ```bash
   ~/.virtualenvs/tap-something/bin/tap-something \
     | ~/.virtualenvs/target-redshift/bin/target-redshift \
       --config ~/singer.io/target_redshift_config.json
   ```

### Config.json

The fields available to be specified in the config file are specified
here.

| Field                       | Type                  | Default    | Details                                                                                                                                                                                                                          |
| --------------------------- | --------------------- | ---------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `redshift_host`             | `["string"]`          | `N/A`      |                                                                                                                                                                                                                                  |
| `redshift_port`             | `["integer", "null"]` | `5432`     |                                                                                                                                                                                                                                  |
| `redshift_database`         | `["string"]`          | `N/A`      |                                                                                                                                                                                                                                  |
| `redshift_username`         | `["string"]`          | `N/A`      |                                                                                                                                                                                                                                  |
| `redshift_password`         | `["string"]`          | `N/A`      |                                                                                                                                                                                                                                  |
| `redshift_schema`           | `["string", "null"]`  | `"public"` |                                                                                                                                                                                                                                  |
| `invalid_records_detect`    | `["boolean", "null"]` | `true`     | Include `false` in your config to disable `target-redshift` from crashing on invalid records                                                                                                                                     |
| `invalid_records_threshold` | `["integer", "null"]` | `0`        | Include a positive value `n` in your config to allow for `target-redshift` to encounter at most `n` invalid records per stream before giving up.                                                                                 |
| `disable_collection`        | `["string", "null"]`  | `false`    | Include `true` in your config to disable [Singer Usage Logging](#usage-logging).                                                                                                                                                 |
| `logging_level`             | `["string", "null"]`  | `"INFO"`   | The level for logging. Set to `DEBUG` to get things like queries executed, timing of those queries, etc. See [Python's Logger Levels](https://docs.python.org/3/library/logging.html#levels) for information about valid values. |
| `max_batch_rows`            | `["integer", "null"]` | `200000`                         | The maximum number of rows to buffer in memory before writing to the destination table in Redshift
| `max_buffer_size`           | `["integer", "null"]` | `104857600` (100MB in bytes)     | The maximum number of bytes to buffer in memory before writing to the destination table in Redshift
| `batch_detection_threshold` | `["integer", "null"]` | `5000`, or 1/40th `max_batch_rows` | How often, in rows received, to count the buffered rows and bytes to check if a flush is necessary. There's a slight performance penalty to checking the buffered records count or bytesize, so this controls how often this is polled in order to mitigate the penalty. This value is usually not necessary to set as the default is dynamically adjusted to check reasonably often.
| `persist_empty_tables`      | `["boolean", "null"]` | `False`    | Whether the Target should create tables which have no records present in Remote.                                                                                                                                                 |
| `default_column_length`     | `["integer", "null"]` | `1000`     | All columns with the VARCHAR(CHARACTER VARYING) type will be have this length if no maxLength is declared in the schema. Range: 1-65535.                                                                                                                                    |
| `default_max_bytes_per_character`     | `["integer", "null"]` | `4`     | Maximum number of bytes per character needed for encoding.  If incoming UTF-16 data has `maxLength: 100` and you set `default_max_bytes_per_character: 2` then the column will be defined as VARCHAR(200) to allow for enough storage. Range: 1-4.                                                                                                                                    |
| `state_support`             | `["boolean", "null"]` | `True`                           | Whether the Target should emit `STATE` messages to stdout for further consumption. In this mode, which is on by default, STATE messages are buffered in memory until all the records that occurred before them are flushed according to the batch flushing schedule the target is configured with.    |
| `target_s3`                 | `["object"]`          | `N/A`      | See `S3` below                                                                                                                                                                                                                   |

#### S3 Config.json

| Field                   | Type                 | Default | Details                                                                      |
| ----------------------- | -------------------- | ------- | ---------------------------------------------------------------------------- |
| `aws_access_key_id`     | `["string"]`         | `N/A`   |                                                                              |
| `aws_secret_access_key` | `["string"]`         | `N/A`   |                                                                              |
| `aws_session_token`     | `["string"]`         | `N/A`   | STS session token if using temporary credentials                             |
| `bucket`                | `["string"]`         | `N/A`   | Bucket where staging files should be uploaded to.                            |
| `key_prefix`            | `["string", "null"]` | `""`    | Prefix for staging file uploads to allow for better delineation of tmp files |

## Known Limitations

- Ignores `STATE` Singer messages.
- Requires a [JSON Schema](https://json-schema.org/) for every stream.
- Only string, string with date-time format, integer, number, boolean,
  object, and array types with or without null are supported. Arrays can
  have any of the other types listed, including objects as types within
  items.
  - Example of JSON Schema types that work
    - `['number']`
    - `['string']`
    - `['string', 'null']`
  - Exmaple of JSON Schema types that **DO NOT** work
    - `['string', 'integer']`
    - `['integer', 'number']`
    - `['any']`
    - `['null']`
- JSON Schema combinations such as `anyOf` and `allOf` are not supported.
- JSON Schema \$ref is partially supported:
  - **_NOTE:_** The following limitations are known to **NOT** fail gracefully
  - Presently you cannot have any circular or recursive `$ref`s
  - `$ref`s must be present within the schema:
    - URI's do not work
    - if the `$ref` is broken, the behaviour is considered unexpected
- Any values which are the `string` `NULL` will be streamed to Redshift as the literal `null`
- Table names are restricted to:
  - 127 characters in length
  - can only be composed of `_`, lowercase letters, numbers, `$`
  - cannot start with `$`
  - ASCII characters
- Field/Column names are restricted to:
  - 127 characters in length
  - ASCII characters
- Fields/Columns are **_ALL_** `nullable`
- Fields/Columns use a default of 1000 characters (in order to be able to work with a large number of columns).

## Usage Logging

[Singer.io](https://www.singer.io/) requires official taps and targets to collect anonymous usage data. This data is only used in aggregate to report on individual tap/targets, as well as the Singer community at-large. IP addresses are recorded to detect unique tap/targets users but not shared with third-parties.

To disable anonymous data collection set `disable_collection` to `true` in the configuration JSON file.

## Developing

`target-redshift` utilizes [setup.py](https://python-packaging.readthedocs.io/en/latest/index.html) for package
management, and [PyTest](https://docs.pytest.org/en/latest/contents.html) for testing.

### Docker

If you have [Docker](https://www.docker.com/) and [Docker Compose](https://docs.docker.com/compose/) installed, you can
easily run the following to get a local env setup quickly.

First, make sure to create a [`.env`](https://docs.docker.com/compose/environment-variables/#the-env-file) file in the root of this repo (it has been `.gitignore`d so don't worry about accidentally staging it).

Therein, fill out the following information:

```sh
REDSHIFT_HOST='<your-host-name>' # Most likely 'localhost'
REDSHIFT_DATABASE='<your-db-name>' # Most likely 'dev'
REDSHIFT_SCHEMA='<your-schema-name>' # Probably 'public'
REDSHIFT_PORT='<your-port>' # Probably 5439
REDSHIFT_USERNAME='<your-user-name'
REDSHIFT_PASSWORD='<your-password>'
DEFAULT_COLUMN_LENGTH='<your-default-column-length>'
DEFAULT_MAX_BYTES_PER_CHARACTER='<your-default-max-bytes-per-character>'
TARGET_S3_AWS_ACCESS_KEY_ID='<AKIA...>'
TARGET_S3_AWS_SECRET_ACCESS_KEY='<secret>'
TARGET_S3_BUCKET='<bucket-string>'
TARGET_S3_KEY_PREFIX='<some-string>' # We use 'target_redshift_test'
```

```sh
$ docker-compose up -d --build
$ docker logs -tf target-redshift_target-redshift_1 # You container names might differ
```

As soon as you see `INFO: Dev environment ready.` you can shell into the container and start running test commands:

```sh
$ docker exec -it target-redshift_target-redshift_1 bash # Your container names might differ
root@... $ source /code/venv--target-redshift/bin/activate
root@... $ pytest
```

See the [PyTest](#pytest) commands below!

### DB

To run the tests, you will need an _actual_ Redshift cluster running, and a user that either:

- Has the ability to _create_ schemas therein

  - This is required if you wish to run multiple versions of the tests, similar
    to how we run our [CI tests by varying the `REDSHIFT_SCHEMA` envvar](https://github.com/datamill-co/target-redshift/blob/master/.circleci/config.yml#L94-L96)

- has access to the `public` schema
  - If the `REDSHIFT_SCHEMA` is seen to be the string `"public"`, the tests will ignore creating and dropping schemas
  - This setup is often preferred for situations in which `GRANT CREATE ON DATABASE db TO user;`
    is viewed as too risky

Make sure to set the following env vars for [PyTest](#pytest):

```sh
$ EXPORT REDSHIFT_HOST='<your-host-name>' # Most likely 'localhost'
$ EXPORT REDSHIFT_DATABASE='<your-db-name>' # Most likely 'dev'
$ EXPORT REDSHIFT_SCHEMA='<your-schema-name>' # Probably 'public'
$ EXPORT REDSHIFT_PORT='<your-port>' # Probably 5439
$ EXPORT REDSHIFT_USERNAME='<your-user-name'
$ EXPORT REDSHIFT_PASSWORD='<your-password>' # Redshift requires passwords
$ EXPORT DEFAULT_COLUMN_LENGTH='<your-default-column-length>'
$ EXPORT DEFAULT_MAX_BYTES_PER_CHARACTER='<your-default-max-bytes-per-character>' # Usually 1 or 2 for European languages
```

### S3

To run the tests, you will need an _actual_ S3 bucket available.

Make sure to set the following env vars for [PyTest](#pytest):

```sh
$ EXPORT TARGET_S3_AWS_ACCESS_KEY_ID='<AKIA...>'
$ EXPORT TARGET_S3_AWS_SECRET_ACCESS_KEY='<secret>'
$ EXPORT TARGET_S3_BUCKET='<bucket-string>'
$ EXPORT TARGET_S3_KEY_PREFIX='<some-string>' # We use 'target_redshift_test'
```

### PyTest

To run tests, try:

```sh
$ python setup.py pytest
```

If you've `bash` shelled into the Docker Compose container ([see above](#docker)), you should be able to simply use:

```sh
$ pytest
```

## Sponsorship

Target Redshift is sponsored by Data Mill (Data Mill Services, LLC) [datamill.co](https://datamill.co/).

Data Mill helps organizations utilize modern data infrastructure and data science to power analytics, products, and services.

---

Copyright Data Mill Services, LLC 2018


CHANGELOG.md
# Changelog

# 0.2.5

- **FEATURES:**
  - [Added the `default_max_bytes_per_character`](https://github.com/datamill-co/target-redshift/pull/9) config option which allows users to declare a default maximum number of bytes to allow per incoming character.  For example, a field in the schema
  declared with `maxLength: 100` with a `default_max_bytes_per_character: 2` will end up as a `varchar(200)` field.  Fields
  with no `maxLength` declared will get `varchar(<default_column_length>)` as before.  `default_max_bytes_per_character` defaults to
  `4` to allow for up to 4 byte UTF and avoid backwards-compatibility issues.

## 0.2.4

- **NOTE:** The `patch` version bump is reflective of the underlying dependency bump from
  [Target-Postgres](https://github.com/datamill-co/target-postgres) signal the
- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.2.1` to `0.2.4`.
    - Performance improvements, bug fixes, and feature support for JSONSchema `anyOf` and `allOf`
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

## 0.2.0

- **NOTE:** The `minor` version bump is not expected to have much effect on folks. This was done to signal the
  output change from the below dependency bump. It is our impression not many are using this feature yet anyways. Since
  this was _not_ a `patch` change, we decided to make this a `minor` instead of `major` change to raise _less_
  concern. Thank you for your patience! Please see the below dependency change for more information.
- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.11` to `0.2.0`.
    - Change of `STATE` message support to be better aligned with the official `singer-target-template`
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.10

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.10` to `0.1.11`.
    - No functional changes impacting `target-redshift`
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.9

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.9` to `0.1.10`.
    - Canonicalization of the `root_table` name to allow for _any_ `stream` name to be passed in
    - Additional testing to ensure safe backwards compatibility
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.8

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.8` to `0.1.9`.
    - Addressed a bug with `ACTIVATE_VERSION` not clearing internal buffers. No _known_ issues arose from this, purely a performance issue
    - Added emitting `STATE` messages when all associated `RECORD` messages have been persisted
    - Added SSL support
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.7

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.7` to `0.1.8`.
    - See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.6

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.6` to `0.1.7`.
    - Fixes a bug with denesting. See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.5

- **DEPENDENCIES:**
  - [Singer Target Postgres](https://pypi.org/project/singer-target-postgres/)
    from `0.1.4` to `0.1.6`.
    - Fixes a number of bugs. See [CHANGELOG](https://github.com/datamill-co/target-postgres/blob/master/CHANGELOG.md)
      for more details.

# 0.0.4

- **FEATURES:**
  - [Added the `persist_empty_tables`](https://github.com/datamill-co/target-postgres) config option detailed in Target-Postgres.

# 0.0.3

- **FEATURES:**
  - [Added the `default_column_length`](https://github.com/datamill-co/target-redshift/pull/9) config option which allows users to declare a default column width for `string` types


tests/test_s3.py
import json
import os

import pytest

from fixtures import CONFIG
from target_redshift.s3 import S3


class Readable:
    def __init__(self, iterable):
        self.iter = iter(iterable)

    def read(self):
        try:
            return json.dumps(next(self.iter)) + '\n'
        except StopIteration:
            return ''


class DownloadableS3(S3):
    def download(self, key):
        self.client.download_file(self.bucket, key, key)

        original = []
        with open(key) as tmp_file:
            for line in tmp_file.readlines():
                original.append(json.loads(line))

        os.remove(key)
        return original


def downloadableS3():
    return DownloadableS3(CONFIG['target_s3']['aws_access_key_id'],
                          CONFIG['target_s3']['aws_secret_access_key'],
                          CONFIG['target_s3']['bucket'],
                          CONFIG['target_s3']['key_prefix'])


def test_persist():
    to_persist = []
    for i in range(100):
        to_persist.append({'a': 123, 'b': 'cdef', 'g': i})

    s3 = downloadableS3()
    bucket, key = s3.persist(Readable(to_persist))
    result = s3.download(key)

    assert to_persist == result


def test_persist__empty():
    s3 = downloadableS3()
    bucket, key = s3.persist(Readable([]))
    result = s3.download(key)

    assert [] == result


target_redshift/s3.py
import uuid

import boto3

SEPARATOR = '__'


class S3:
    def __init__(
        self,
        aws_access_key_id,
        aws_secret_access_key,
        bucket,
        key_prefix='',
        aws_session_token=None
    ):
        self._credentials = {'aws_access_key_id': aws_access_key_id,
                             'aws_secret_access_key': aws_secret_access_key,
                             'aws_session_token': aws_session_token}
        self.client = boto3.client(
            's3',
            aws_access_key_id=aws_access_key_id,
            aws_secret_access_key=aws_secret_access_key,
            aws_session_token=aws_session_token)
        self.bucket = bucket
        self.key_prefix = key_prefix

    def credentials(self):
        return self._credentials

    def persist(self, readable, key_prefix=''):
        key = self.key_prefix + (key_prefix or '') + str(uuid.uuid4()).replace('-', '')

        self.client.upload_fileobj(
            _EncodeBinaryReadable(readable),
            self.bucket,
            key)

        return [self.bucket, key]


class _EncodeBinaryReadable:
    def __init__(self, readable_obj):
        self.input = readable_obj

    def readable(self):
        return True

    def read(self, *args, **kwargs):
        if len(args) > 0:
            max_bytes = args[0]
        else:
            max_bytes = None
        output = b''
        while (max_bytes is not None and len(output) < max_bytes) or True:  ## TODO: overflow?
            line = self.input.read()
            if line == '':
                return output
            output += line.encode('utf-8')
        return output


target_redshift/__init__.py
import psycopg2
import singer
from singer import utils
from target_postgres import target_tools
from target_postgres.postgres import MillisLoggingConnection

from target_redshift.redshift import RedshiftTarget
from target_redshift.s3 import S3

LOGGER = singer.get_logger()

REQUIRED_CONFIG_KEYS = [
    'redshift_host',
    'redshift_database',
    'redshift_username',
    'redshift_password',
    'target_s3'
]


def main(config, input_stream=None):
    with psycopg2.connect(
            connection_factory=MillisLoggingConnection,
            host=config.get('redshift_host'),
            port=config.get('redshift_port', 5439),
            dbname=config.get('redshift_database'),
            user=config.get('redshift_username'),
            password=config.get('redshift_password')
    ) as connection:
        s3_config = config.get('target_s3')
        s3 = S3(s3_config.get('aws_access_key_id'),
                s3_config.get('aws_secret_access_key'),
                s3_config.get('bucket'),
                s3_config.get('key_prefix'),
                aws_session_token=s3_config.get('aws_session_token'))

        redshift_target = RedshiftTarget(
            connection,
            s3,
            redshift_schema=config.get('redshift_schema', 'public'),
            logging_level=config.get('logging_level'),
            default_column_length=config.get('default_column_length', 1000),
            default_max_bytes_per_character=config.get('default_max_bytes_per_character', 4),
            persist_empty_tables=config.get('persist_empty_tables')
        )

        if input_stream:
            target_tools.stream_to_target(input_stream, redshift_target, config=config)
        else:
            target_tools.main(redshift_target)


def cli():
    args = utils.parse_args(REQUIRED_CONFIG_KEYS)

    main(args.config)


target_redshift/redshift.py
from copy import deepcopy
import re

from psycopg2 import sql
from target_postgres import json_schema
from target_postgres.postgres import PostgresError, PostgresTarget, RESERVED_NULL_DEFAULT
from target_postgres.singer_stream import (
    SINGER_LEVEL
)
from target_postgres.sql_base import SEPARATOR


class RedshiftError(PostgresError):
    """
    Raise this when there is an error with regards to Redshift streaming
    """


def _make_schema_nullable(schema):
    # Redshift does not allow for creation of columns which are non null without a default
    nullable_schema = deepcopy(schema)
    nullable_properties = nullable_schema['properties']

    for field in nullable_properties:
        nullable_properties[field] = json_schema.make_nullable(nullable_properties[field])

        if 'anyOf' in nullable_properties[field]:
            nullable_properties[field]['anyOf'] = [
                json_schema.make_nullable(sub_schema)
                for sub_schema in nullable_properties[field]['anyOf']
            ]

    return nullable_schema


class RedshiftTarget(PostgresTarget):
    """
    Placeholder for specific Redshift implementation of a Singer Target.
    """

    # https://docs.aws.amazon.com/redshift/latest/dg/r_names.html
    IDENTIFIER_FIELD_LENGTH = 127
    DEFAULT_COLUMN_LENGTH = 1000
    DEFAULT_MAX_BYTES_PER_CHARACTER = 4

    MAX_VARCHAR = 65535
    CREATE_TABLE_INITIAL_COLUMN = '_sdc_target_redshift_create_table_placeholder'
    CREATE_TABLE_INITIAL_COLUMN_TYPE = 'BOOLEAN'

    def __init__(self, connection, s3, *args,
        redshift_schema='public',
        logging_level=None,
        default_column_length=DEFAULT_COLUMN_LENGTH,
        default_max_bytes_per_character=DEFAULT_MAX_BYTES_PER_CHARACTER,
        persist_empty_tables=False,
        **kwargs):

        self.LOGGER.info(
            'RedshiftTarget created with established connection: `{}`, schema: `{}`'.format(connection.dsn,
                                                                                            redshift_schema))

        self.s3 = s3
        self.default_column_length = default_column_length
        self.default_max_bytes_per_character = default_max_bytes_per_character
        PostgresTarget.__init__(self, connection, postgres_schema=redshift_schema, logging_level=logging_level,
                                persist_empty_tables=persist_empty_tables, add_upsert_indexes=False)

    def write_batch(self, stream_buffer):
        # WARNING: Using mutability here as there's no simple way to copy the necessary data over
        self.LOGGER.info('write_batch: Schema before nullability: {}'.format(stream_buffer.schema))
        nullable_stream_buffer = stream_buffer
        nullable_stream_buffer.schema = _make_schema_nullable(stream_buffer.schema)

        self.LOGGER.info('write_batch: Schema after nullability: {}'.format(stream_buffer.schema))

        return PostgresTarget.write_batch(self, nullable_stream_buffer)

    def upsert_table_helper(self, connection, table_schema, metadata, log_schema_changes=True):
        self.LOGGER.info('upsert_table_helper: Schema before nullability: {}'.format(table_schema))

        nullable_table_schema = deepcopy(table_schema)
        nullable_table_schema['schema'] = _make_schema_nullable(nullable_table_schema['schema'])
        self.LOGGER.info('upsert_table_helper: Schema after nullability: {}'.format(nullable_table_schema))
        return PostgresTarget.upsert_table_helper(self,
                                                  connection,
                                                  nullable_table_schema,
                                                  metadata,
                                                  log_schema_changes=log_schema_changes)

    def add_table(self, cur, path, name, metadata):
        self._validate_identifier(name)

        create_table_sql = sql.SQL('CREATE TABLE {}.{} ({} {})').format(
            sql.Identifier(self.postgres_schema),
            sql.Identifier(name),
            # Redshift does not allow for creation of tables with no columns
            sql.Identifier(self.CREATE_TABLE_INITIAL_COLUMN),
            sql.SQL(self.CREATE_TABLE_INITIAL_COLUMN_TYPE))

        cur.execute(sql.SQL('{};').format(
            create_table_sql))

        self._set_table_metadata(cur, name, {'path': path,
                                             'version': metadata.get('version', None),
                                             'schema_version': metadata['schema_version']})

        self.add_column_mapping(cur,
                                name,
                                (self.CREATE_TABLE_INITIAL_COLUMN,),
                                self.CREATE_TABLE_INITIAL_COLUMN,
                                json_schema.make_nullable({'type': json_schema.BOOLEAN}))

    def sql_type_to_json_schema(self, sql_type, is_nullable):
        if sql_type == 'character varying':
            schema = {'type': [json_schema.STRING]}
            if is_nullable:
                return json_schema.make_nullable(schema)
            return schema

        return PostgresTarget.sql_type_to_json_schema(self, sql_type, is_nullable)

    def json_schema_to_sql_type(self, schema):
        psql_type = PostgresTarget.json_schema_to_sql_type(self, schema)

        assert not 'NOT NULL' in psql_type, 'Redshift does not support `NOT NULL` without a default. Got: {}'.format(
            psql_type,
            schema
        )

        max_length = schema.get('maxLength')
        if max_length is None:
            max_length == self.default_column_length
        else:
            max_length *= self.default_max_bytes_per_character

        if max_length > self.MAX_VARCHAR:
            max_length = self.MAX_VARCHAR

        if psql_type.upper() == 'TEXT':
            return 'varchar({})'.format(max_length)

        return psql_type

    def add_column(self, cur, table_name, column_name, column_schema):
        self.LOGGER.info('add_column({}, {}, {})'.format(
            table_name, column_name, column_schema
        ))
        PostgresTarget.add_column(self, cur, table_name, column_name, column_schema)


    def persist_csv_rows(self,
                         cur,
                         remote_schema,
                         temp_table_name,
                         columns,
                         csv_rows):
        key_prefix = temp_table_name + SEPARATOR

        bucket, key = self.s3.persist(csv_rows,
                                      key_prefix=key_prefix)

        credentials = self.s3.credentials()
        aws_access_key_id = credentials.get('aws_access_key_id')
        aws_secret_access_key= credentials.get('aws_secret_access_key')
        aws_session_token = credentials.get('aws_session_token')

        copy_sql = sql.SQL('COPY {}.{} ({}) FROM {} CREDENTIALS {} FORMAT AS CSV NULL AS {}').format(
            sql.Identifier(self.postgres_schema),
            sql.Identifier(temp_table_name),
            sql.SQL(', ').join(map(sql.Identifier, columns)),
            sql.Literal('s3://{}/{}'.format(bucket, key)),
            sql.Literal('aws_access_key_id={};aws_secret_access_key={}{}'.format(
                aws_access_key_id,
                aws_secret_access_key,
                ";token={}".format(aws_session_token) if aws_session_token else '',
            )),
            sql.Literal(RESERVED_NULL_DEFAULT))

        cur.execute(copy_sql)

        pattern = re.compile(SINGER_LEVEL.format('[0-9]+'))
        subkeys = list(filter(lambda header: re.match(pattern, header) is not None, columns))

        update_sql = self._get_update_sql(remote_schema['name'],
                                          temp_table_name,
                                          remote_schema['key_properties'],
                                          columns,
                                          subkeys)

        cur.execute(update_sql)
